python/lsst/ap/association/diaCalculationPlugins.py

```diff
@@ -0,0 +1,158 @@
+# This file is part of ap_association.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import numpy as np
+
+import lsst.geom as geom
+
+from .diaCalculation import (
+    DiaObjectCalculationPluginConfig,
+    DiaObjectCalculationPlugin)
+from lsst.meas.base.pluginRegistry import register
+
+__all__ = ("MeanDiaPositionConfig", "MeanDiaPosition",
+           "WeightedMeanDiaPsFluxConfig", "WeightedMeanDiaPsFlux")
+
+
+class MeanDiaPositionConfig(DiaObjectCalculationPluginConfig):
+    pass
+
+
+@register("ap_meanPosition")
+class MeanDiaPosition(DiaObjectCalculationPlugin):
+    """Compute the mean position of a DiaObject given a set of DiaSources.
+    """
+
+    ConfigClass = MeanDiaPositionConfig
+
+    @classmethod
+    def getExecutionOrder(cls):
+        return cls.DEFAULT_CATALOGCALCULATION
+
+    def __init__(self, config, name, metadata):
+        DiaObjectCalculationPlugin.__init__(self, config, name, metadata)
+
+    def calculate(self, diaObject, diaSources, **kwargs):
+        """Compute the mean ra/dec position of the diaObject given the
+        diaSource locations.
+
+        Parameters
+        ----------
+        diaObject : `dict`
+            Summary object to store values in.
```

Comment:
I'm worried about the semantics of this. Specifically, I wonder if we ought to be stricter about a schema for the `diaObject`.

Consider two cases:

1. I write my own `JohnsMeanDiaPosition` plugin, which also calculates `ra` and `decl` values and writes them to the `diaObject` dict. Whichever plugin runs _last_ will “win”, in that my values might over-write yours (or vice versa) in the dict, and we've no way of knowing which results we actually get.

2. A dependent plugin (like `DependentDiaPlugin`, below) has to “guess” what field names were used by the earlier plugins which ran.

In the basic CatalogCalculation case, we avoid these issues by editing the schema up-front — the task will fail to init if plugins are using duplicate names. Since you don't have a schema here, that doesn't work.

There are a couple of things I'd suggest:

1. Be really clear about documenting the schema and semantics which should be used in the `diaObject` dict. For example, do all the keys map directly to the field names used in the DPDD? If so, say so. If there's some other convention being used to define the field names, say that.

2. The plugin should check that nothing already in the dict is being overwritten. For example, something like:

```python
class DiaObjectCalculationPlugin(object):
    def calculate(self, diaObject, diaSources, **kwargs):
        diaObjectCopy = diaObject.copy()
        results = self.runCalculation(diaObjectCopy, diaSources, **kwargs)
        for k, v in results:
            if k in diaObject:
                self.fail(...)  # or raise? Not sure.
            diaObject[k] = v
          
    def runCalculation(...):
        # (do stuff)
        return {calculated_value=1, calculated_other_value=2}
```
(I think you'd want to tweak that code a bit, rather than use it as is.)

Reply:
It might make more sense to enforce this at pluggin creation. As in, make sure that no pluggin would try to write to the same place twice based on it's output. Likewise, we could make sure using the same check to see if columns exist that a column a pluggin requires to run is created by a previous pluggin. Sound good?

Also, yes, the columns of the DiaObject are required to be the same as that of the Ppdb/DPDD.

Reply:
Added new class variables defining input and output columns for a given plugin and tests within DiaCalculationTask that validate that these don't collide with other plugins and have the proper output columns for plugins that are dependent on previous runs.

Reply:
This looks good, but as we discussed please check that https://github.com/lsst/ap_association/blob/c3c5b8f2b79474ac678d6c842b539e8a47d80978/python/lsst/ap/association/diaCalculation.py#L220 really does sort plugins by execution order.

Reply:
Added a unittest that gives the dependent plugin before the independent plugin in the config. A `ValueError` is not raises so the plugins are indeed sorted by execution order.